File: sdk-python/examples/test_sdk_sync.py
#!/usr/bin/env python3
"""
SDK synchronization test — verifies all features added during SDK sync.

Tests: workspace CRUD + StorageType, file ops (move, copy, get_file_info, exists),
sandbox (exists, wait_for_state), process (shell, exec), error handling.

Usage:
    python examples/test_sdk_sync.py [--server ADDR] [--apikey KEY] [--image IMAGE]

Flags:
    --server  gRPC server address (default: localhost:9090)
    --apikey  gRPC API key or JWT (optional)
    --image   Sandbox container image (default: workspace-test:latest)
"""

import argparse
import os
import sys

# Add parent directory to path for development
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from workspace_sdk import (
    WorkspaceClient,
    CreateWorkspaceParams,
    CreateSandboxParams,
    SandboxState,
    ProcessError,
    NotFoundError,
)


class TestRunner:
    def __init__(self):
        self.results: list[tuple[str, bool, Exception | None]] = []

    def run(self, name: str, fn):
        print(f"\n── {name} ──")
        try:
            fn()
            self.results.append((name, True, None))
            print("  PASSED")
        except Exception as e:
            self.results.append((name, False, e))
            print(f"  FAILED: {e}")

    def summary(self):
        print("\n╔══════════════════════════════════════════════════╗")
        print("║                  Test Summary                    ║")
        print("╚══════════════════════════════════════════════════╝")
        passed = failed = 0
        for name, ok, err in self.results:
            status = "✓ PASS" if ok else "✗ FAIL"
            if ok:
                passed += 1
            else:
                failed += 1
            print(f"  {status}  {name}")
            if err:
                print(f"         Error: {err}")
        print(f"\n  Total: {passed} passed, {failed} failed")
        if failed > 0:
            sys.exit(1)


def test_workspace_crud(client: WorkspaceClient):
    """Test 1: Workspace CRUD + StorageType"""
    # Create
    ws = client.workspace.create(
        CreateWorkspaceParams(name="py-sync-test", metadata={"test": "sync"})
    )
    print(f"  Created workspace: {ws.id} (storage_type={ws.storage_type})")

    try:
        # Verify default storage type is managed
        if ws.storage_type != "managed":
            raise AssertionError(
                f"expected storage_type=managed, got {ws.storage_type}"
            )

        # Get
        got = client.workspace.get(ws.id)
        if got.name != "py-sync-test":
            raise AssertionError(f"name mismatch: got {got.name}")

        # List
        ws_list = client.workspace.list()
        found = any(w.id == ws.id for w in ws_list)
        if not found:
            raise AssertionError("workspace not found in list")

        print("  Workspace CRUD + StorageType OK")
    finally:
        client.workspace.delete(ws.id)


def test_file_ops(client: WorkspaceClient):
    """Test 2: All workspace file operations"""
    ws = client.workspace.create(CreateWorkspaceParams(name="py-file-ops-test"))

    try:
        ws_id = ws.id

        # WriteFile + ReadFile
        client.workspace.write_file(ws_id, "hello.txt", "Hello World")
        content = client.workspace.read_file(ws_id, "hello.txt")
        if content != "Hello World":
            raise AssertionError(f"content mismatch: got {content!r}")
        print("  WriteFile + ReadFile OK")

        # Mkdir + ListFiles
        client.workspace.mkdir(ws_id, "subdir")
        files = client.workspace.list_files(ws_id, ".")
        print(f"  ListFiles: {len(files)} items")

        # GetFileInfo
        info = client.workspace.get_file_info(ws_id, "hello.txt")
        if info.type != "file" or info.name != "hello.txt":
            raise AssertionError(
                f"file info mismatch: name={info.name} type={info.type}"
            )
        print(f"  GetFileInfo: name={info.name} type={info.type} size={info.size}")

        # FileExists (via exists)
        exists = client.workspace.exists(ws_id, "hello.txt")
        if not exists:
            raise AssertionError("file should exist")

        exists = client.workspace.exists(ws_id, "no-such-file.txt")
        if exists:
            raise AssertionError("non-existent file should not exist")
        print("  FileExists OK")

        # CopyFile
        client.workspace.copy_file(ws_id, "hello.txt", "hello_copy.txt")
        copy_content = client.workspace.read_file(ws_id, "hello_copy.txt")
        if copy_content != "Hello World":
            raise AssertionError(f"copy content mismatch: got {copy_content!r}")
        print("  CopyFile OK")

        # MoveFile
        client.workspace.move_file(ws_id, "hello_copy.txt", "hello_moved.txt")
        exists = client.workspace.exists(ws_id, "hello_copy.txt")
        if exists:
            raise AssertionError("moved source should not exist")
        moved_content = client.workspace.read_file(ws_id, "hello_moved.txt")
        if moved_content != "Hello World":
            raise AssertionError("moved content mismatch")
        print("  MoveFile OK")

        # DeleteFile
        client.workspace.delete_file(ws_id, "hello_moved.txt", recursive=False)
        exists = client.workspace.exists(ws_id, "hello_moved.txt")
        if exists:
            raise AssertionError("deleted file should not exist")
        print("  DeleteFile OK")
    finally:
        client.workspace.delete(ws.id)


def test_sandbox_features(client: WorkspaceClient, image: str):
    """Test 3: Sandbox Exists + WaitForState"""
    ws = client.workspace.create(CreateWorkspaceParams(name="py-sandbox-test"))

    try:
        try:
            sb = client.sandbox.create(
                CreateSandboxParams(
                    workspace_id=ws.id,
                    template=image,
                    name="py-sync-test-sandbox",
                )
            )
        except Exception as e:
            print(f"  Sandbox creation failed (Docker may not be available): {e}")
            print("  Skipping sandbox-dependent tests")
            return

        try:
            print(f"  Created sandbox: {sb.id} (state: {sb.state})")

            # Exists
            exists = client.sandbox.exists(sb.id)
            if not exists:
                raise AssertionError("sandbox should exist")

            exists = client.sandbox.exists("non-existent-sandbox")
            if exists:
                raise AssertionError("non-existent sandbox should not exist")
            print("  Sandbox.exists OK")

            # WaitForState
            sb = client.sandbox.wait_for_state(
                sb.id, SandboxState.RUNNING, timeout=30.0
            )
            print(f"  WaitForState: reached {sb.state}")
        finally:
            client.sandbox.delete(sb.id, force=True)
    finally:
        client.workspace.delete(ws.id)


def test_process_features(client: WorkspaceClient, image: str):
    """Test 4: Process Shell + Exec"""
    ws = client.workspace.create(CreateWorkspaceParams(name="py-process-test"))

    try:
        try:
            sb = client.sandbox.create(
                CreateSandboxParams(workspace_id=ws.id, template=image)
            )
        except Exception as e:
            print(f"  Sandbox creation failed: {e}")
            print("  Skipping process tests")
            return

        try:
            client.sandbox.wait_for_state(sb.id, SandboxState.RUNNING, timeout=30.0)

            # Shell
            result = client.process.shell(sb.id, "echo hello && echo world")
            if result.exit_code != 0:
                raise AssertionError(f"shell exit code: {result.exit_code}")
            print(f"  Shell output: {result.stdout}", end="")
            print("  Shell OK")

            # Exec
            stdout = client.process.exec(sb.id, "echo", "Python", "SDK", "exec")
            print(f"  Exec output: {stdout}", end="")
            print("  Exec OK")
        finally:
            client.sandbox.delete(sb.id, force=True)
    finally:
        client.workspace.delete(ws.id)


def test_error_handling(client: WorkspaceClient, image: str):
    """Test 5: Error Handling (ProcessError, NotFound)"""
    ws = client.workspace.create(CreateWorkspaceParams(name="py-error-test"))

    try:
        try:
            sb = client.sandbox.create(
                CreateSandboxParams(workspace_id=ws.id, template=image)
            )
        except Exception as e:
            print("  Skipping error handling tests (no Docker)")
            return

        try:
            client.sandbox.wait_for_state(sb.id, SandboxState.RUNNING, timeout=30.0)

            # ProcessError from exec
            try:
                client.process.exec(sb.id, "false")
                raise AssertionError("exec 'false' should fail")
            except ProcessError as pe:
                print(
                    f"  ProcessError caught: sandbox={pe.sandbox_id} command={pe.command}"
                )
            except AssertionError:
                raise
            except Exception as e:
                raise AssertionError(f"expected ProcessError, got {type(e).__name__}: {e}")

            # NotFound error
            try:
                client.workspace.get("non-existent-ws")
                raise AssertionError("get non-existent workspace should fail")
            except NotFoundError:
                print("  NotFoundError OK")
            except AssertionError:
                raise
            except Exception as e:
                raise AssertionError(f"expected NotFoundError, got {type(e).__name__}: {e}")
        finally:
            client.sandbox.delete(sb.id, force=True)
    finally:
        client.workspace.delete(ws.id)


def main():
    parser = argparse.ArgumentParser(description="Python SDK Sync Test")
    parser.add_argument("--server", default="localhost:9090", help="gRPC server address")
    parser.add_argument("--apikey", default="", help="gRPC API key or JWT")
    parser.add_argument(
        "--image", default="workspace-test:latest", help="Sandbox container image"
    )
    args = parser.parse_args()

    print("╔══════════════════════════════════════════════════╗")
    print("║  Python SDK Sync Test — Verify All SDK Features ║")
    print("╚══════════════════════════════════════════════════╝")
    print(f"  Server: {args.server}\n")

    runner = TestRunner()

    with WorkspaceClient(args.server, api_key=args.apikey, timeout=60.0) as client:
        runner.run("1. Workspace CRUD + StorageType", lambda: test_workspace_crud(client))
        runner.run("2. Workspace File Operations (all)", lambda: test_file_ops(client))
        runner.run(
            "3. Sandbox Exists + WaitForState",
            lambda: test_sandbox_features(client, args.image),
        )
        runner.run(
            "4. Process Shell + Exec",
            lambda: test_process_features(client, args.image),
        )
        runner.run(
            "5. Error Handling (ProcessError, NotFound)",
            lambda: test_error_handling(client, args.image),
        )

    runner.summary()


if __name__ == "__main__":
    main()

File: sdk-typescript/examples/test_sdk_sync.ts
/**
 * SDK synchronization test — verifies all features added during SDK sync.
 *
 * Tests: workspace CRUD + StorageType, file ops (move, copy, getFileInfo, exists),
 * sandbox (exists, waitForState), process (shell, exec), error handling.
 *
 * Usage:
 *   npx ts-node examples/test_sdk_sync.ts [options]
 *
 * Options:
 *   --server <addr>   gRPC server address (default: localhost:9090)
 *   --apikey <key>    gRPC API key or JWT (optional)
 *   --image <image>   Sandbox container image (default: workspace-test:latest)
 */

import { WorkspaceClient } from '../src/client';
import { ProcessError, WorkspaceError } from '../src/errors';
import type { Workspace, Sandbox } from '../src/types';

// Parse command line arguments
function parseArgs(): { server: string; apikey: string; image: string } {
  const args = process.argv.slice(2);
  let server = 'localhost:9090';
  let apikey = '';
  let image = 'workspace-test:latest';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--server' && args[i + 1]) {
      server = args[++i];
    } else if (args[i] === '--apikey' && args[i + 1]) {
      apikey = args[++i];
    } else if (args[i] === '--image' && args[i + 1]) {
      image = args[++i];
    }
  }

  return { server, apikey, image };
}

interface TestResult {
  name: string;
  passed: boolean;
  error?: Error;
}

const results: TestResult[] = [];

async function runTest(name: string, fn: () => Promise<void>): Promise<void> {
  console.log(`\n── ${name} ──`);
  try {
    await fn();
    results.push({ name, passed: true });
    console.log('  PASSED');
  } catch (err) {
    results.push({ name, passed: false, error: err as Error });
    console.log(`  FAILED: ${(err as Error).message}`);
  }
}

async function testWorkspaceCRUD(client: WorkspaceClient): Promise<void> {
  // Create
  const ws = await client.workspace.create({
    name: 'ts-sync-test',
    metadata: { test: 'sync' },
  });
  console.log(`  Created workspace: ${ws.id} (storageType=${ws.storageType})`);

  try {
    // Verify default storage type is managed
    if (ws.storageType !== 'managed') {
      throw new Error(`expected storageType=managed, got ${ws.storageType}`);
    }

    // Get
    const got = await client.workspace.get(ws.id);
    if (got.name !== 'ts-sync-test') {
      throw new Error(`name mismatch: got ${got.name}`);
    }

    // List
    const list = await client.workspace.list();
    const found = list.some((w) => w.id === ws.id);
    if (!found) {
      throw new Error('workspace not found in list');
    }

    console.log('  Workspace CRUD + StorageType OK');
  } finally {
    await client.workspace.delete(ws.id);
  }
}

async function testFileOps(client: WorkspaceClient): Promise<void> {
  const ws = await client.workspace.create({ name: 'ts-file-ops-test' });

  try {
    const wsId = ws.id;

    // WriteFile + ReadFile
    await client.workspace.writeFile(wsId, 'hello.txt', 'Hello World');
    const content = await client.workspace.readFile(wsId, 'hello.txt');
    if (content !== 'Hello World') {
      throw new Error(`content mismatch: got "${content}"`);
    }
    console.log('  WriteFile + ReadFile OK');

    // Mkdir + ListFiles
    await client.workspace.mkdir(wsId, 'subdir');
    const files = await client.workspace.listFiles(wsId, '.');
    console.log(`  ListFiles: ${files.length} items`);

    // GetFileInfo
    const info = await client.workspace.getFileInfo(wsId, 'hello.txt');
    if (info.type !== 'file' || info.name !== 'hello.txt') {
      throw new Error(`file info mismatch: name=${info.name} type=${info.type}`);
    }
    console.log(`  GetFileInfo: name=${info.name} type=${info.type} size=${info.size}`);

    // FileExists
    let exists = await client.workspace.fileExists(wsId, 'hello.txt');
    if (!exists) {
      throw new Error('file should exist');
    }

    exists = await client.workspace.fileExists(wsId, 'no-such-file.txt');
    if (exists) {
      throw new Error('non-existent file should not exist');
    }
    console.log('  FileExists OK');

    // CopyFile
    await client.workspace.copyFile(wsId, 'hello.txt', 'hello_copy.txt');
    const copyContent = await client.workspace.readFile(wsId, 'hello_copy.txt');
    if (copyContent !== 'Hello World') {
      throw new Error(`copy content mismatch: got "${copyContent}"`);
    }
    console.log('  CopyFile OK');

    // MoveFile
    await client.workspace.moveFile(wsId, 'hello_copy.txt', 'hello_moved.txt');
    exists = await client.workspace.fileExists(wsId, 'hello_copy.txt');
    if (exists) {
      throw new Error('moved source should not exist');
    }
    const movedContent = await client.workspace.readFile(wsId, 'hello_moved.txt');
    if (movedContent !== 'Hello World') {
      throw new Error('moved content mismatch');
    }
    console.log('  MoveFile OK');

    // DeleteFile
    await client.workspace.deleteFile(wsId, 'hello_moved.txt', false);
    exists = await client.workspace.fileExists(wsId, 'hello_moved.txt');
    if (exists) {
      throw new Error('deleted file should not exist');
    }
    console.log('  DeleteFile OK');
  } finally {
    await client.workspace.delete(ws.id);
  }
}

async function testSandboxFeatures(client: WorkspaceClient, image: string): Promise<void> {
  const ws = await client.workspace.create({ name: 'ts-sandbox-test' });

  try {
    let sb: Sandbox;
    try {
      sb = await client.sandbox.create({
        workspaceId: ws.id,
        template: image,
        name: 'ts-sync-test-sandbox',
      });
    } catch (e) {
      console.log(`  Sandbox creation failed (Docker may not be available): ${(e as Error).message}`);
      console.log('  Skipping sandbox-dependent tests');
      return;
    }

    try {
      console.log(`  Created sandbox: ${sb.id} (state: ${sb.state})`);

      // Exists
      let exists = await client.sandbox.exists(sb.id);
      if (!exists) {
        throw new Error('sandbox should exist');
      }

      exists = await client.sandbox.exists('non-existent-sandbox');
      if (exists) {
        throw new Error('non-existent sandbox should not exist');
      }
      console.log('  Sandbox.exists OK');

      // WaitForState
      sb = await client.sandbox.waitForState(sb.id, 'running', AbortSignal.timeout(30000));
      console.log(`  WaitForState: reached ${sb.state}`);
    } finally {
      await client.sandbox.delete(sb.id, true);
    }
  } finally {
    await client.workspace.delete(ws.id);
  }
}

async function testProcessFeatures(client: WorkspaceClient, image: string): Promise<void> {
  const ws = await client.workspace.create({ name: 'ts-process-test' });

  try {
    let sb: Sandbox;
    try {
      sb = await client.sandbox.create({
        workspaceId: ws.id,
        template: image,
      });
    } catch (e) {
      console.log(`  Sandbox creation failed: ${(e as Error).message}`);
      console.log('  Skipping process tests');
      return;
    }

    try {
      await client.sandbox.waitForState(sb.id, 'running', AbortSignal.timeout(30000));

      // Shell
      const result = await client.process.shell(sb.id, 'echo hello && echo world');
      if (result.exitCode !== 0) {
        throw new Error(`shell exit code: ${result.exitCode}`);
      }
      process.stdout.write(`  Shell output: ${result.stdout}`);
      console.log('  Shell OK');

      // Exec
      const stdout = await client.process.exec(sb.id, 'echo', 'TypeScript', 'SDK', 'exec');
      process.stdout.write(`  Exec output: ${stdout}`);
      console.log('  Exec OK');
    } finally {
      await client.sandbox.delete(sb.id, true);
    }
  } finally {
    await client.workspace.delete(ws.id);
  }
}

async function testErrorHandling(client: WorkspaceClient, image: string): Promise<void> {
  const ws = await client.workspace.create({ name: 'ts-error-test' });

  try {
    let sb: Sandbox;
    try {
      sb = await client.sandbox.create({
        workspaceId: ws.id,
        template: image,
      });
    } catch (e) {
      console.log('  Skipping error handling tests (no Docker)');
      return;
    }

    try {
      await client.sandbox.waitForState(sb.id, 'running', AbortSignal.timeout(30000));

      // ProcessError from exec
      try {
        await client.process.exec(sb.id, 'false');
        throw new Error("exec 'false' should fail");
      } catch (err) {
        if (err instanceof ProcessError) {
          console.log(`  ProcessError caught: sandbox=${err.sandboxId} command=${err.command}`);
        } else if ((err as Error).message === "exec 'false' should fail") {
          throw err;
        } else {
          throw new Error(`expected ProcessError, got ${(err as Error).constructor.name}: ${(err as Error).message}`);
        }
      }

      // NotFound error
      try {
        await client.workspace.get('non-existent-ws');
        throw new Error('get non-existent workspace should fail');
      } catch (err) {
        if (err instanceof WorkspaceError && err.message.toLowerCase().includes('not found')) {
          console.log('  NotFoundError OK');
        } else if ((err as Error).message === 'get non-existent workspace should fail') {
          throw err;
        } else {
          throw new Error(`expected NotFound, got: ${(err as Error).message}`);
        }
      }
    } finally {
      await client.sandbox.delete(sb.id, true);
    }
  } finally {
    await client.workspace.delete(ws.id);
  }
}

async function main() {
  const { server, apikey, image } = parseArgs();

  console.log('╔══════════════════════════════════════════════════╗');
  console.log('║   TS SDK Sync Test — Verify All SDK Features    ║');
  console.log('╚══════════════════════════════════════════════════╝');
  console.log(`  Server: ${server}\n`);

  const clientOpts: { apiKey?: string } = {};
  if (apikey) clientOpts.apiKey = apikey;
  const client = new WorkspaceClient(server, clientOpts);

  try {
    await runTest('1. Workspace CRUD + StorageType', () => testWorkspaceCRUD(client));
    await runTest('2. Workspace File Operations (all)', () => testFileOps(client));
    await runTest('3. Sandbox Exists + WaitForState', () => testSandboxFeatures(client, image));
    await runTest('4. Process Shell + Exec', () => testProcessFeatures(client, image));
    await runTest('5. Error Handling (ProcessError, NotFound)', () => testErrorHandling(client, image));
  } finally {
    client.close();
  }

  // Print summary
  console.log('\n╔══════════════════════════════════════════════════╗');
  console.log('║                  Test Summary                    ║');
  console.log('╚══════════════════════════════════════════════════╝');
  let passed = 0;
  let failed = 0;
  for (const r of results) {
    const status = r.passed ? '✓ PASS' : '✗ FAIL';
    if (r.passed) {
      passed++;
    } else {
      failed++;
    }
    console.log(`  ${status}  ${r.name}`);
    if (r.error) {
      console.log(`         Error: ${r.error.message}`);
    }
  }
  console.log(`\n  Total: ${passed} passed, ${failed} failed`);
  if (failed > 0) {
    process.exit(1);
  }
}

main().catch((e) => {
  console.error('Error:', e);
  process.exit(1);
});
